bundle/run/job_args.go

```diff
@@ -0,0 +1,193 @@
+	return nil
+}
+
+func (a jobTaskNotebookParamArgs) CompleteArgs(args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	parameters := make(map[string]string)
+	for _, t := range a.Tasks {
+		if nt := t.NotebookTask; nt != nil {
+			maps.Copy(parameters, nt.BaseParameters)
+		}
+	}
+	return genericCompleteKeyValueArgs(args, toComplete, maps.Keys(parameters))
+}
+
+type jobTaskJarParamArgs struct {
+	*resources.Job
+}
+
+func (a jobTaskJarParamArgs) ParseArgs(args []string, opts *Options) error {
+	opts.Job.jarParams = append(opts.Job.jarParams, args...)
+	return nil
+}
+
+func (a jobTaskJarParamArgs) CompleteArgs(args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return nil, cobra.ShellCompDirectiveNoFileComp
+}
+
+type jobTaskPythonParamArgs struct {
+	*resources.Job
+}
+
+func (a jobTaskPythonParamArgs) ParseArgs(args []string, opts *Options) error {
+	opts.Job.pythonParams = append(opts.Job.pythonParams, args...)
+	return nil
+}
+
+func (a jobTaskPythonParamArgs) CompleteArgs(args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return nil, cobra.ShellCompDirectiveNoFileComp
+}
+
+type jobTaskSparkSubmitParamArgs struct {
+	*resources.Job
+}
+
+func (a jobTaskSparkSubmitParamArgs) ParseArgs(args []string, opts *Options) error {
+	opts.Job.sparkSubmitParams = append(opts.Job.pythonParams, args...)
+	return nil
+}
+
+func (a jobTaskSparkSubmitParamArgs) CompleteArgs(args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return nil, cobra.ShellCompDirectiveNoFileComp
+}
+
+type jobTaskType int
+
+const (
+	jobTaskTypeNotebook jobTaskType = iota + 1
+	jobTaskTypeSparkJar
+	jobTaskTypeSparkPython
+	jobTaskTypeSparkSubmit
+	jobTaskTypePipeline
+	jobTaskTypePythonWheel
+	jobTaskTypeSql
+	jobTaskTypeDbt
+	jobTaskTypeRunJob
+)
+
+func (r *jobRunner) posArgsHandler() argsHandler {
+	job := r.job
+	if job == nil || job.JobSettings == nil {
+		return nopArgsHandler{}
+	}
+
+	// Handle job parameters, if any are defined.
+	if len(job.Parameters) > 0 {
+		return &jobParameterArgs{job}
+	}
+
+	// Handle task parameters otherwise.
+	var seen = make(map[jobTaskType]bool)
+	var typ jobTaskType
+	for _, t := range job.Tasks {
+		if t.NotebookTask != nil {
+			typ = jobTaskTypeNotebook
+			seen[typ] = true
+		}
+		if t.SparkJarTask != nil {
+			typ = jobTaskTypeSparkJar
+			seen[typ] = true
+		}
+		if t.SparkPythonTask != nil {
+			typ = jobTaskTypeSparkPython
+			seen[typ] = true
+		}
+		if t.SparkSubmitTask != nil {
+			typ = jobTaskTypeSparkSubmit
+			seen[typ] = true
+		}
+		if t.PipelineTask != nil {
+			typ = jobTaskTypePipeline
+			seen[typ] = true
+		}
+		if t.PythonWheelTask != nil {
+			typ = jobTaskTypePythonWheel
+			seen[typ] = true
+		}
+		if t.SqlTask != nil {
+			typ = jobTaskTypeSql
+			seen[typ] = true
+		}
+		if t.DbtTask != nil {
+			typ = jobTaskTypeDbt
+			seen[typ] = true
+		}
+		if t.RunJobTask != nil {
+			typ = jobTaskTypeRunJob
+			seen[typ] = true
+		}
```

Comment:
I like the convenience here but I'm a bit worried about maintainability. We should expect there to be many more new task types over the coming years. Maybe we can do something smart with a unit test and reflection to make sure we cover all cases? (GPT4 can help write a test like that)

Reply:
> This loop is relevant only for tasks that accept task parameters and not job parameters. I expect all new task types to exclusively work with job parameters.

I'm not sure we can be certain about that yet. There might be other tasks like the pipeline task type with special options.

> Also, the nopArgsHandler returns an error if args are specified and we don't know how to handle them.

Does that mean we get new customer bug reports when we introduce new task types? I like failing loudly, but that also doesn't seem ideal?